### PR TITLE
Fix delay processor by delaying onError and onComplete signals

### PR DESCRIPTION
**Pull request by marcantoinefortier**
## 📖 Description

🐛 Bug fix (non-breaking change which fixes an issue)

💔 Breaking change (fix or feature that would cause existing functionality to change)

The previous delay processor implementation was not respecting the reactive spec for that operator, since the onError and onComplete signals also need to be delayed.

## 💭 Motivation and Context
Since our .debounce() implementation recently introduced a breaking change with publishers' completion, we need to have .delay() operator with the proper behaviour.

## 🧪 How Has This Been Tested?
This is unit tested and has been tested locally in a client's application.

## 🦀 Dispatch

- `#dispatch/kmp`

### CODE REVIEW

File: streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/processors/DelayProcessor.kt [hunk 1]
```diff
@@ -25,20 +27,41 @@ class DelayProcessor<T>(
         private val timerFactory: TimerFactory
     ) : ProcessorSubscription<T, T>(subscriber) {
 
+        private val serialQueue = SynchronousSerialQueue()
         private val cancellableManagerProvider = CancellableManager()
 
+        private var blocksToRun: List<() -> Unit> by atomic(emptyList())
```
Comment (npresseault) [inline streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/processors/DelayProcessor.kt:33]: Maybe we could use AtomicListReference and we can use it as a regular mutable list.

**Review by npresseault — APPROVED** [review]

File: streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/processors/DelayProcessor.kt (hunk 1 above)
Comment (marcantoinefortier) [reply streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/processors/DelayProcessor.kt:33]: Good suggestion I din't know this existed. 👍🏻 

Howerver, I tried using it but there's a few missing operators on that AtomicListReference, such as `.clear()` and `.drop(n)` for example, which would make it easier to read.

For `.clear()` I could use `blocksToRun.removeAll(blocksToRun.value)`, but this does not seem too much eficient. Same thing for `.drop(n)` where I could do the following :

```
blocksToRun.value.firstOrNull()?.let { block ->
    block.invoke()
    blocksToRun.remove(block)
}
```

What do you think about keeping the `atomic(emptyList())` and updating it aftwerards once we implement these operators on the AtomicListReference?

Comment (marcantoinefortier) [reply streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/processors/DelayProcessor.kt:33]: @npresseault Merging this for now as we need a quick (and clean) aternative to the latest `.debounce()` breaking change. I'll address the AtomicListReference issue later if needed. Thanks

Comment (npresseault) [reply streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/processors/DelayProcessor.kt:33]: Yes, works for me. Didn't notice not everything was implemented with AtomicListReference

**Pull request merged**